samples/net/zstream/big_http_download/src/big_http_download.c

```diff
@@ -0,0 +1,285 @@
+		if (out_len < 0) {
+			return out_len;
+		}
+		buf = (const char *)buf + out_len;
+		len -= out_len;
+	}
+
+	return 0;
+}
+
+int skip_headers(struct zstream *stream)
+{
+	int state = 0;
+
+	while (1) {
+		char c;
+		int st;
+
+		st = zstream_read(stream, &c, 1);
+		if (st <= 0) {
+			return st;
+		}
+
+		if (state == 0 && c == '\r') {
+			state++;
+		} else if (state == 1 && c == '\n') {
+			state++;
+		} else if (state == 2 && c == '\r') {
+			state++;
+		} else if (state == 3 && c == '\n') {
+			break;
+		} else {
+			state = 0;
+		}
+	}
+
+	return 1;
+}
+
+void print_hex(const unsigned char *p, int len)
+{
+	while (len--) {
+		printf("%02x", *p++);
+	}
+}
+
+void download(struct addrinfo *ai, mbedtls_ssl_config *tls_conf)
+{
+	int sock;
+	struct zstream_sock stream_sock;
+	struct zstream_tls stream_tls;
+	struct zstream *stream;
+
+	cur_bytes = 0;
+
+	sock = socket(ai->ai_family, ai->ai_socktype, ai->ai_protocol);
+	CHECK(sock);
+	printf("sock = %d\n", sock);
+	CHECK(connect(sock, ai->ai_addr, ai->ai_addrlen));
+
+	zstream_sock_init(&stream_sock, sock);
+	stream = (struct zstream *)&stream_sock;
+
+	if (tls_conf) {
+		zstream_tls_init(&stream_tls, stream, tls_conf, host);
+		stream = (struct zstream *)&stream_tls;
+	}
+
+	sendall(stream, "GET /", SSTRLEN("GET /"));
+	sendall(stream, uri_path, strlen(uri_path));
+	sendall(stream, " HTTP/1.0\r\n", SSTRLEN(" HTTP/1.0\r\n"));
+	sendall(stream, "Host: ", SSTRLEN("Host: "));
+	sendall(stream, host, strlen(host));
+	sendall(stream, "\r\n\r\n", SSTRLEN("\r\n\r\n"));
+	zstream_flush(stream);
+
+	if (skip_headers(stream) <= 0) {
+		printf("EOF or error in response headers\n");
+		goto error;
+	}
+
+	mbedtls_md_starts(&hash_ctx);
+
+	while (1) {
+		int len = zstream_read(stream, response, sizeof(response) - 1);
+
+		if (len < 0) {
+			printf("Error reading response\n");
+			goto error;
+		}
+
+		if (len == 0) {
+			break;
+		}
+
+		mbedtls_md_update(&hash_ctx, response, len);
+
+		cur_bytes += len;
+		printf("%u bytes\r", cur_bytes);
+
+		response[len] = 0;
+		/*printf("%s\n", response);*/
+	}
+
+	printf("\n");
+
+	mbedtls_md_finish(&hash_ctx, response_hash);
+
+	printf("Hash: ");
+	print_hex(response_hash, mbedtls_md_get_size(hash_info));
+	printf("\n");
+
+	if (memcmp(response_hash, download_hash,
+		   mbedtls_md_get_size(hash_info)) != 0) {
+		printf("HASH MISMATCH!\n");
+	}
+
+error:
+	zstream_close(stream);
+}
+
+int main(void)
+{
+	static struct addrinfo hints;
+	struct addrinfo *res;
+	int st;
+	char *p;
+	unsigned int total_bytes = 0;
+	mbedtls_ssl_config *tls_conf = NULL;
+
+	setbuf(stdout, NULL);
+
+	if (strncmp(download_url, "http://", SSTRLEN("http://")) == 0) {
+		port = "80";
+		p = download_url + SSTRLEN("http://");
+	} else if (strncmp(download_url, "https://",
+		   SSTRLEN("https://")) == 0) {
+		if (ztls_get_tls_client_conf(&tls_conf) < 0) {
+			printf("Unable to initialize TLS\n");
+			return 1;
+		}
+		mbedtls_ssl_conf_authmode(tls_conf, MBEDTLS_SSL_VERIFY_NONE);
```

Comment:
Since this is the currently "canonical" sample for using this API, I would really like to see what an application that requires a fully verified certificate chain looks like.

Reply:
So, TLS support requires covering 2 areas: a) TLS communication; b) TLS configuration. Zstream concentrates on solving first part, based on the idea that regardless of configuration details, the communication would work the same. Configuration is actually handled by separate sub-API, whose discussion is started in https://github.com/zephyrproject-rtos/zephyr/issues/6086 . *Beginnings* of the implementation of this API are included in commit "tls_conf: Convenience API for TLS configuration settings", file  include/net/tls_conf.h, direct link working until next rebase is: https://github.com/zephyrproject-rtos/zephyr/pull/5985/commits/dd619df53798daf133c31ddfa6cd647cd9053ec2 .

As can be seen, currently it includes just 4 functions, which give an idea how it's supposed to work. A complete configuration API, with support for fully verified chain and other TLS features, would include dozens of such functions. The API also likely will need restructuring to support TLS offloading. However, the work done so far and this PR was based on the specific aim - provide socket-based analog of net_app's TLS support. The latter doesn't specifically include support for "fully verified certificate chain", and neither this PR does.

I'll be happy to work with all the interested parties to define a good TLS configuration API - if this PR is selected as the way forward. And indeed, defining such an API satisfying all the parties (security team on one side, vendors with TLS offloading implementations on the other side, and open-source Zephyr RTOS in the middle) would require a lot of discussion and effort, why it wasn't done so far.

Reply:
The net-app based echo_client sample appears to show how to use that API and also to verify certificates. Am I misunderstanding?

I think we should have a fully worked example for a real-world situation where at least the server cert is verified with both this API and the setsockopt() API. Doing it with a "real" website also is important. This is what users are going to have to do to connect to the popular clouds.

Several people have suggested using libtls as a source of inspiration, presumably because it has a nice API. In the introductory talk about it linked to in #5900, you will see that the authors insist on not merging new API without real-world application code that uses it. Seems smart to me, since there are apparently so many "annoying parts".

Reply:
> The net-app based echo_client sample appears to show how to use that API and also to verify certificates. Am I misunderstanding?

Well, the best answer I have - "depends". Verifying certificates is trivial - you load root certificates, period. The rest is done automagically by mbedTLS. (You don't even need to enabled verification, it's on by default. You disable it as an extra step if you want that.) And there're samples for mbedTLS API how to do that - sure, there's no rocket science.

But "real-world situation" for TLS means "e.g. to work as a web browser". Which can connect to (almost) any TLS site in the world and verify it - if browser is fresh enough. But that means few hundreds of KBs of root store. And the whole exercise becomes a logistical and organizational one, not technical. E.g., which boards will be able to run such an example? (qemu_x86 and frdm_k64f, that's all?) What source to use for these root certs? How to update them (also easily)? Because if done now and forgotten, in a a year someone will cry "you have insecure example with outdated root certificates!11". Etc, etc. no problems to resolve *right now*, IMHO.

Reply:
> Several people have suggested using libtls as a source of inspiration, presumably because it has a nice API.

I gave my analysis of that proverbial libtls here: https://github.com/zephyrproject-rtos/zephyr/issues/6086#issuecomment-365311559

Summary: tls_conf and libtls are the same things, modulo the differences between them and us: a) they have terabytes of diskspace, gigabytes of RAM, megabytes of stack to spend, we don't; b) they're all about only TLS, we're about commonality between TCP and TLS. In all other aspects we're going to have as nice as them, just in a bit different way, adapted for resources we have.

Extending response above, there're not even dozens, but hundreds of TLS options to configure (counting e.g. each ciphersuite as a "separate option"). No, I did not consider how to handle each of such options, so don't want to pretend I did (especially that it requires involvement of other parties). Just wrapping mbedTLS API is the baseline:

~~~
#define ztls_conf_foo mbedtls_conf_foo
~~~

We'll need to to do more due diligence on that of course (unlikely will work for offloading for example), the current beginnings are where I pointed to.

Reply:
> We're not building web browsers here. This is an RTOS.
> I think "real world" for many (most?) connected embedded

You see, we even disagree what "real world example" is. For you, it's actually some kind of artificial example, for me, it's something, what even my grandma can appreciate ;-). That's the reason such a thing is not included yet.

As you know, @d3zd3z is currently working on "full stack" (not "real world") example on connecting to google cloud and verifying its certificates (using net_app, and apparently raw mbedtls API). If my approach is selected, I'll use his feedback as the input how to support stuff. While still making an oath that my API will support "grandma" usecase too. I.e., if someone will want to make a connected device showing to a grandma a quote of the day fetched via old good REST API in the fully secure way - they will be able to do that with Zephyr ;-).

Reply:
> I gave my analysis of that proverbial libtls here: #6086 (comment)

I know. I've read the entire backlog of all of the issues and pull requests you've called to our attention on this issue. You are missing my point.

> Extending response above, there're not even dozens, but hundreds of TLS options to configure (counting e.g. each ciphersuite as a "separate option"). No, I did not consider how to handle each of such options

I'm not asking for that. I'm asking for a single, complete example of an application program which uses your API and follows best practices for using TLS, which includes verifying the server cert. That leaves an incredible amount of latitude to pick the server and a particular set of values for the many tunables out there to suit a single use case, which is what the appplication development users of any TLS API are going to have to do.

Reply:
> I know. I've read the entire backlog of all of the issues and pull requests you've called to our attention on this issue.

Thanks ;-).

> You are missing my point.

I'd hope I don't, just trying to show difference in points of view. As I explained, the matter of verifying peer cert is a matter of loading root cert(s). Now you're saying that this particular option is the most important of all. But for me, there's a hundred of options, all important, all to be supported. But we can't support them all at once, and sustained importance of any particular option is just 1% ;-).

> That leaves an incredible amount of latitude to pick the server and a particular set

So, I can't implement what you ask until you specify the server to use. Then I'll need to do bunch of legwork of figuring out and converting its root cert chain. Then it will break after some time (when they change cert). Feel free to specify the server name, though in the situation, I can't promise when it'll be done ;-).

Reply:
> I fail to see how an example which supports the minimum requirements to connect to AWS's MQTT servers via TLS is in any way "artificial".

My point is that we should shoot for "maximum requirements", and the "default" example should show them, as used "in wide real world", assessable by wide masses of developers. From that, we can scale down to "tiny world". That's not a contradiction to your view, just different prioritization.

Reply:
> Now you're saying that this particular option is the most important of all. But for me, there's a hundred of options, all important, all to be supported.

People are entitled to their opinions, of course, but it seems to me that a handshake, followed by client verification of the server certificate, followed by exchange of session data, is pretty much TLS 101.

For context, I have reached out to @d3zd3z to work with him / benefit from his efforts on MQTTS, and I am evaluating both TLS APIs under consideration from that user perspective. This was among the first questions I had when looking at both pull requests. It's not an arbitrary idea.